kroxylicious/src/main/java/io/kroxylicious/proxy/bootstrap/FilterChainFactory.java

```diff
@@ -7,46 +7,91 @@
-import java.util.stream.Collectors;
+import java.util.concurrent.ScheduledExecutorService;
 
 import io.kroxylicious.proxy.config.FilterDefinition;
+import io.kroxylicious.proxy.config.PluginFactory;
+import io.kroxylicious.proxy.config.PluginFactoryRegistry;
+import io.kroxylicious.proxy.filter.Filter;
 import io.kroxylicious.proxy.filter.FilterAndInvoker;
-import io.kroxylicious.proxy.filter.FilterCreationContext;
-import io.kroxylicious.proxy.filter.InvalidFilterConfigurationException;
-import io.kroxylicious.proxy.internal.filter.NettyFilterContext;
-import io.kroxylicious.proxy.service.FilterFactoryManager;
+import io.kroxylicious.proxy.filter.FilterFactory;
+import io.kroxylicious.proxy.filter.FilterFactoryContext;
+import io.kroxylicious.proxy.plugin.PluginConfigurationException;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
 
 /**
  * Abstracts the creation of a chain of filter instances, hiding the configuration
  * required for instantiation at the point at which instances are created.
  * New instances are created during initialization of a downstream channel.
  */
 public class FilterChainFactory {
-    private final List<FilterDefinition> filterDefinitions;
 
-    public FilterChainFactory(List<FilterDefinition> filterDefinitions) {
-        validateFilterDefinitions(filterDefinitions);
-        this.filterDefinitions = Optional.ofNullable(filterDefinitions).orElse(List.of());
+    record UninitializedFilterFactory(String instanceName, FilterFactory<? super Object, ? super Object> filterFactory, Object config) {
+
+        private InitializedFilterFactory initialize(FilterFactoryContext context) {
+            Object result;
+            try {
+                result = filterFactory.initialize(context, config);
+            }
+            catch (Exception e) {
+                throw new PluginConfigurationException("Exception initializing filter factory " + instanceName + " with config " + config + ": " + e.getMessage(), e);
+            }
+            return new InitializedFilterFactory(filterFactory, result);
+        }
+    }
+
+    record InitializedFilterFactory(FilterFactory<? super Object, ? super Object> filterFactory, Object initResult) {
+        public Filter create(FilterFactoryContext context) {
+            try {
+                return filterFactory().createFilter(context, initResult);
+            }
+            catch (Exception e) {
+                throw new PluginConfigurationException("Exception instantiating filter using factory " + filterFactory, e);
+            }
+        }
     }
 
     /**
-     * Validate that a collection of FilterDefinitions contains all the required configuration to configure the filter.
-     * @param filterDefinitions the candidates to validate.
-     * @throws InvalidFilterConfigurationException If there are problems.
+     * A constant Class type with non-raw {@code FilterFactory} type parameter,
+     * for the avoidance of raw type warnings.
```

Comment:
You still add `@SuppressWarnings({ "unchecked", "rawtypes" })` so is it worth it?

Reply:
The difference is having the suppressed warning in a single place, where the suppression is very specific. In contrast to possibly having to annotate multiple use sites, where the suppression might accidentally apply to some other expressions. But I'm also not entirely convinced that it's worth it in this case, since there's only a single production use site.

Reply:
Actually, I forgot that Java now lets us annotate local variables, so it's really not worth it.